annotatedcsv/decode.go

```diff
@@ -0,0 +1,451 @@
+		if ftype == nil {
+			continue
+		}
+		colType, ok := columnTypes[col.Type]
+		if !ok {
+			// ignore invalid type and use string
+			colType = stringCol
+		}
+		fkind := fieldKindOf(ftype)
+		convert, ok := conversions[conv{colType, fkind}]
+		if !ok {
+			return nil, fmt.Errorf("cannot convert from column type %s to %v", col.Type, ftype)
+		}
+		setters[i] = convert
+	}
+	return setters, nil
+}
+
+// toInterface returns a function for setting value of given column type
+// to an interface{} field
+func toInterface(col colType) fieldSetter {
+	t := canonicalTypes[col]
+	convert, ok := conversions[conv{col, fieldKindOf(t)}]
+	if !ok {
+		panic("conversion not found (should not happen)")
+	}
+	return func(v reflect.Value, s string) error {
+		e := reflect.New(t).Elem()
+		if err := convert(e, s); err != nil {
+			return err
+		}
+		v.Set(e)
+		return nil
+	}
+}
+
+// toString sets a string to a value of type string
+func toString(v reflect.Value, s string) error {
+	v.SetString(s)
+	return nil
+}
+
+// toFloat converts a string to a value of type float
+func toFloat(v reflect.Value, s string) error {
+	x, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		return err
+	}
+	if v.OverflowFloat(x) {
+		return fmt.Errorf("overflow")
+	}
+	v.SetFloat(x)
+	return nil
+}
+
+// toBool converts a string to a value of type bool
+func toBool(v reflect.Value, s string) error {
+	var b bool
+	switch s {
+	case "true":
+		b = true
+	case "false":
+		b = false
+	default:
+		return fmt.Errorf("invalid bool value")
+	}
+	v.SetBool(b)
+	return nil
+}
+
+// toInt converts a string to a value of type signed int
+func toInt(v reflect.Value, s string) error {
+	x, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return err
+	}
+	if v.OverflowInt(x) {
+		return fmt.Errorf("overflow")
+	}
+	v.SetInt(x)
+	return nil
+}
+
+// toUint converts a string to a value of type unsigned int
+func toUint(v reflect.Value, s string) error {
+	x, err := strconv.ParseUint(s, 10, 64)
+	if err != nil {
+		return err
+	}
+	if v.OverflowUint(x) {
+		return fmt.Errorf("overflow")
+	}
+	v.SetUint(x)
+	return nil
+}
+
+// toTime converts a string to a time value
+func toTime(v reflect.Value, s string) error {
+	x, err := time.Parse(time.RFC3339Nano, s)
+	if err != nil {
+		return err
+	}
+	v.Set(reflect.ValueOf(x))
+	return nil
+}
+
+// toDuration converts a string to a duration value
+func toDuration(v reflect.Value, s string) error {
+	x, err := time.ParseDuration(s)
+	if err != nil {
+		return err
+	}
+	v.Set(reflect.ValueOf(x))
+	return nil
+}
+
+// toBytes decodes a base64 encoded string to a slice of bytes
+func toBytes(v reflect.Value, s string) error {
+	x, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return err
+	}
+	v.Set(reflect.ValueOf(x))
+	return nil
+}
+
+// stringTernary returns second argument of first is empty, otherwise first
+func stringTernary(s string, d string) string {
+	if s != "" {
+		return s
+	}
+	return d
+}
+
+func init() {
```

Comment:
FWIW I tend to prefer to avoid `init` functions like this in favour of a function initializer style:

```
var conversions = func() map[conv] fieldSetter {
    etc
}()
```

That way you never have to worry about initialization order.

Reply:
I was trying this, but as `toInterface{}` needs `conversions`, it's a cycle dependency. So I've chosen `init` as a solution

Reply:
Good point! I hadn't realised that. Maybe add a comment noting that?

Reply:
Do you mean a comment which explains why it is initialized in the `init()`?